### PR TITLE
Implement Deterministic Faulty Graph Builder

### PR DESCRIPTION
**Pull request by danieldgt**
Now, is possible to generate graph with all combinations of possible fault links

Commit:Add

### CODE REVIEW

File: src/util/Combinatorics.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+                if ((m & (1 << i)) != 0) {
+                    aCombination.add(arrayObjects[i]);
+                }
+            }
+            result.add(aCombination);
+        }
+        return result;
+    }
+
+    public static List<Set> allCombinationsOf(Set objects, int maxSize) {
```
Comment (jmferreiratech) [inline src/util/Combinatorics.java]: You're not using this method, therefore you can remove it.

File: src/util/Combinatorics.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package util;
+
+import java.util.*;
+
+public class Combinatorics {
+
+    public static List<Set> allCombinationsOf(Set objects) {
```
Comment (jmferreiratech) [inline src/util/Combinatorics.java]: Could it be made a [generic method](https://docs.oracle.com/javase/tutorial/extra/generics/methods.html)?

File: src/util/DeterministicFaultyGraphBuilder.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    private static int maxFaultyOfGraph(int rows, int columns) {
+        return numberOfEdges(rows, columns) - numberOfMinGoodEdges(rows, columns);
+    }
+
+    private static int numberOfGraphsForNumberFaulty(int rows, int columns, int nmFaulty) {
+        int n = numberOfEdges(rows, columns);
+        int k = nmFaulty;
+        Integer result = (int) CombinatoricsUtils.binomialCoefficient(n, k);
```
Comment (jmferreiratech) [inline src/util/DeterministicFaultyGraphBuilder.java]: That dependency should be hidden at 'util.Combinatorics'. Better than that we could implement it.

File: src/util/DeterministicFaultyGraphBuilder.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        int maxFaultyOfGraph = maxFaultyOfGraph(rows, columns);
+        int nmTopology = 0;
+        for (int i = 0; i <= maxFaultyOfGraph; i++) {
+            nmTopology += numberOfGraphsForNumberFaulty(rows, columns, i);
+        }
+
+        return nmTopology;
+    }
+
+}
```
Comment (jmferreiratech) [inline src/util/DeterministicFaultyGraphBuilder.java]: Always place a newline before EOF.

File: src/util/DeterministicFaultyGraphBuilder.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
+package util;
+
+import org.apache.commons.math3.util.CombinatoricsUtils;
+
+import java.util.*;
+
+public class DeterministicFaultyGraphBuilder {
+
+    static public Graph generateGraph(int rows, int columns, int index) {
```
Comment (jmferreiratech) [inline src/util/DeterministicFaultyGraphBuilder.java:7]: Please break this method. It's way too big.

File: src/util/DeterministicFaultyGraphBuilder.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    private static int numberOfGraphsForNumberFaulty(int rows, int columns, int nmFaulty) {
+        int n = numberOfEdges(rows, columns);
+        int k = nmFaulty;
+        Integer result = (int) CombinatoricsUtils.binomialCoefficient(n, k);
+
+        return result;
+    }
+
+    private static int numberOfTopology(int rows, int columns) {
+
```
Comment (jmferreiratech) [inline src/util/DeterministicFaultyGraphBuilder.java]: Please check how we use vertical indent on our code base.

**Review by jmferreiratech — CHANGES_REQUESTED** [review]

File: src/util/Combinatorics.java (hunk 2 above)
Comment (jmferreiratech) [reply src/util/Combinatorics.java]: And please, make the constructor of Combinatorics private.

Comment (danieldgt) [reply src/util/Combinatorics.java]: when you say, generic method, I may do that:

public static List<Set<Object>> allCombinationsOf(Set<? extends Object> objects) {
        List<Set<Object>> result = new ArrayList<Set<Object>>();

or 

public static List<Set> allCombinationsOf(Set<? extends Object> objects) {
.....

File: src/util/Combinatorics.java (hunk 1 above)
Comment (danieldgt) [reply src/util/Combinatorics.java]: ok, I forgot

**Pull request closed without merging**